### PR TITLE
[batch2] add memory request to docker container config

### PR DESCRIPTION
**Pull request by jigold**
This is an enhancement of #7498.

We adjust the number of cores in a user's request to make sure they have enough memory based on the memory per core rather than giving them the exact resources they asked for.

### CODE REVIEW

**Comment by jigold** [comment]
Closing this while I debug with dev deploy.

**Comment by jigold** [comment]
This should be ready for review.

**Comment by jigold** [comment]
Do I need to add resource requests to the build.yaml file and CI build.py? I'm worried about things getting out of memory errors now.

**Comment by jigold** [comment]
I can't figure out why my out of memory test isn't working. It's reporting exit code 0 and no out of memory error even though when I do the same thing locally on my computer with docker run or on an instance in the cluster, I get exit code 137 and out of memory. I'm limiting the docker run command to the same amount of bytes that the docker command in the worker should be limiting it to (looked at the docker output in the worker logs). I think the next step is to try using curl to run docker containers rather than the docker cli.

File: batch2/batch/front_end/front_end.py [hunk 1]
```diff
@@ -283,7 +285,20 @@
...
                     resources['memory'] = BATCH_JOB_DEFAULT_MEMORY
 
-                cores_mcpu = parse_cpu_in_mcpu(resources['cpu'])
+                req_cores_mcpu = parse_cpu_in_mcpu(resources['cpu'])
+                req_memory_bytes = parse_memory_in_bytes(resources['memory'])
+
+                if req_cores_mcpu == 0:
+                    raise web.HTTPBadRequest(
+                        reason=f'resource requests for job {id} are unsatisfiable: '
+                        f'cpu must be greater than 0')
```
Comment (cseed) [inline batch2/batch/front_end/front_end.py]: 0 seems satisfiable, we just don't allow it.

File: batch2/batch/front_end/front_end.py [hunk 2]
```diff
@@ -283,7 +285,20 @@
...
+
+                cores_mcpu = adjust_cores_for_memory_request(req_cores_mcpu, req_memory_bytes, WORKER_TYPE)
+
+                if cores_mcpu > WORKER_CORES * 1000:
+                    raise web.HTTPBadRequest(
+                        reason=f'resource requests for job {id} are unsatisfiable: '
+                        f'cpu={resources["cpu"]}, memory={resources["memory"]}')
```
Comment (cseed) [inline batch2/batch/front_end/front_end.py]: Might be good to include the maximums.

File: batch2/batch/utils.py [hunk 3]
```diff
@@ -17,6 +18,48 @@ def parse_cpu_in_mcpu(cpu_string):
...
+
+def worker_memory_per_core_bytes(worker_type):
+    if worker_type == 'standard':
+        m = 3.75
+    elif worker_type == 'highmem':
+        m = 6.5
+    else:
+        assert worker_type == 'highcpu', worker_type
+        m = 0.9
+    return int(0.9 * m * 1000**3)  # GCE memory/core are in GB not GiB
```
Comment (cseed) [inline batch2/batch/utils.py]: I think the 0.9 stuff is overkill.  Can remove it?  This means converting 3.375 => 3.75, too, in a few spots.

File: batch2/batch/worker.py [hunk 4]
```diff
@@ -160,13 +168,17 @@ def step(self, name, **kwargs):
         status = {
             'state': cstate['Status'],
             'started_at': cstate['StartedAt'],
-            'finished_at': cstate['FinishedAt']
+            'finished_at': cstate['FinishedAt'],
```
Comment (cseed) [inline batch2/batch/worker.py]: Unnecessary.

File: batch2/batch/worker.py [hunk 5]
```diff
@@ -270,8 +282,9 @@ def step(self, name, **kwargs):
     #     state: str,
     #     started_at: str, (date)
     #     finished_at: str, (date)
+    #     out_of_memory: boolean, (optional)
     #     error: str, (one of error, exit_code will be present)
-    #     exit_code: int
+    #     exit_code: int,
```
Comment (cseed) [inline batch2/batch/worker.py]: ?

File: batch2/batch/worker.py [hunk 6]
```diff
@@ -424,7 +439,7 @@ def job_id(self):
 
     @property
     def id(self):
-        return (self.batch_id, self.job_id)
+        return self.batch_id, self.job_id
```
Comment (cseed) [inline batch2/batch/worker.py]: I'm not sure if we're consistent about this, but I prefer parens when forming tuples.

File: batch2/test/test_batch.py [hunk 7]
```diff
@@ -81,6 +81,29 @@ def test_bad_command(self):
         assert j._get_error(status, 'main') is not None
         assert status['state'] == 'Error', status
 
+    def test_invalid_resource_requests(self):
+        builder = self.client.create_batch()
+        resources = {'cpu': '1', 'memory': '28Gi'}
+        builder.create_job('ubuntu:18.04', ['true'], resources=resources)
+        with self.assertRaisesRegex(aiohttp.client.ClientResponseError, 'resource requests'):
```
Comment (cseed) [inline batch2/test/test_batch.py]: I'd prefer `'resource requests.*unsatisfiable'`

File: batch2/test/test_batch.py [hunk 8]
```diff
@@ -81,6 +81,29 @@ def test_bad_command(self):
...
+        resources = {'cpu': '0', 'memory': '1Gi'}
+        builder.create_job('ubuntu:18.04', ['true'], resources=resources)
+        with self.assertRaisesRegex(aiohttp.client.ClientResponseError, 'resource requests'):
+            builder.submit()
+
+    def test_out_of_memory(self):
+        builder = self.client.create_batch()
+        resources = {'cpu': '0.1', 'memory': '10M'}
+        j = builder.create_job('python:3.6-slim-stretch',
+                               ['python', '-c', 'x = "a" * 400 * 1000**2; print(x)'],
```
Comment (cseed) [inline batch2/test/test_batch.py]: Do you need the print?  That could create some giant logs.

File: batch2/batch/worker.py [hunk 9]
```diff
@@ -160,13 +168,17 @@ def step(self, name, **kwargs):
...
+            'finished_at': cstate['FinishedAt'],
         }
         cerror = cstate['Error']
         if cerror:
             status['error'] = cerror
         else:
             status['exit_code'] = cstate['ExitCode']
+
+        if cstate['OOMKilled']:
+            status['out_of_memory'] = True
```
Comment (cseed) [inline batch2/batch/worker.py]: Nice.  We might set this unconditionally, with False if it didn't run out of memory.  What's the docker container state in this case?  The (Hail) container state?  Are they error?

Comment (jigold) [reply batch2/batch/worker.py]: The docker container was exiting with exit code 137. No error. I don't think the logs said it was out of memory error, so I think you have to look at the flag to know docker killed it for out of memory. The Hail container state will be failed.

Comment (jigold) [reply batch2/batch/worker.py]: Let me know if you think this field shouldn't be optional.

File: batch2/batch/front_end/front_end.py [hunk 10]
```diff
@@ -283,7 +285,21 @@
...
+                        reason=f'resource requests for job {id} are not allowed: '
+                        f'cpu must be greater than 0')
+
+                cores_mcpu = adjust_cores_for_memory_request(req_cores_mcpu, req_memory_bytes, WORKER_TYPE)
+
+                if cores_mcpu > WORKER_CORES * 1000:
+                    raise web.HTTPBadRequest(
+                        reason=f'resource requests for job {id} are unsatisfiable: '
+                        f'requested: cpu={resources["cpu"]}, memory={resources["memory"]} '
+                        f'maximum: cpu={WORKER_CORES}, memory={worker_memory_per_core_gb(WORKER_TYPE)}GB')
```
Comment (cseed) [inline batch2/batch/front_end/front_end.py]: You want the total memory, not the memory per cores.

File: batch2/batch/front_end/front_end.py (hunk 1 above)
Comment (cseed) [reply batch2/batch/front_end/front_end.py]: Sorry to nit, this still doesn't read well.  Resource requests are always allowed, just that these ones are bad.  How about: "bad resource request for job {id}: cpu cannot be 0".

File: batch2/batch/utils.py [hunk 11]
```diff
@@ -17,6 +18,53 @@ def parse_cpu_in_mcpu(cpu_string):
...
+    return m
+
+
+def worker_memory_per_core_bytes(worker_type):
+    m = worker_memory_per_core_gb(worker_type)
+    return int(m * 1000**3)  # GCE memory/core are in GB not GiB
+
+
+def memory_bytes_to_cores_mcpu(memory_in_bytes, worker_type):
+    return int((memory_in_bytes / worker_memory_per_core_bytes(worker_type)) * 1000)
```
Comment (cseed) [inline batch2/batch/utils.py]: To make sure the requested memory is satisfied, I think we should round up here: use `math.ceil`.

File: batch2/batch/worker.py (hunk 9 above)
Comment (cseed) [reply batch2/batch/worker.py]: I think it shouldn't be optional.

I'm debating about whether we should return this as failed, exit_code=137 or error, error=Out Of Memory Exception.  out_of_memory is good, it preserves the knowledge of out of memory, but, after years of trying to read the Spark tea leaves, it wouldn't be clearer to users as a proper error message (with exit_code=137).

Let's leave it as you have it now and see if people get confused.  We'll definitely want to document this at some point.  We don't want to leave people to read the tea leaves.

File: batch2/batch/front_end/front_end.py [hunk 12]
```diff
@@ -41,7 +42,6 @@
 REQUEST_TIME = pc.Summary('batch2_request_latency_seconds', 'Batch request latency in seconds', ['endpoint', 'verb'])
 REQUEST_TIME_GET_JOBS = REQUEST_TIME.labels(endpoint='/api/v1alpha/batches/batch_id/jobs', verb="GET")
 REQUEST_TIME_GET_JOB = REQUEST_TIME.labels(endpoint='/api/v1alpha/batches/batch_id/jobs/job_id', verb="GET")
-REQUEST_TIME_GET_JOBS = REQUEST_TIME.labels(endpoint='/api/v1alpha/batches/batch_id/jobs', verb="GET")
```
Comment (cseed) [inline batch2/batch/front_end/front_end.py]: Why is this getting deleted?  That method is still there, isn't it?

Comment (jigold) [reply batch2/batch/front_end/front_end.py]: It's the same exact line as the line two lines above it.

Comment (cseed) [reply batch2/batch/front_end/front_end.py]: Oh, good catch!

File: batch2/batch/front_end/front_end.py [hunk 13]
```diff
@@ -283,7 +285,22 @@
...
+
+                if req_cores_mcpu == 0:
+                    raise web.HTTPBadRequest(
+                        reason=f'bad resource request for job {id}: '
+                        f'cpu cannot be 0')
+
+                cores_mcpu = adjust_cores_for_memory_request(req_cores_mcpu, req_memory_bytes, WORKER_TYPE)
+
+                if cores_mcpu > WORKER_CORES * 1000:
+                    total_memory_available = round(worker_memory_per_core_gb(WORKER_TYPE) * WORKER_CORES, 2)
```
Comment (cseed) [inline batch2/batch/front_end/front_end.py]: I don't think we should round.  The memory per core in GB is generally a nice, round number, and if you round, you're going to be able to generate messages like: unsatisfiable: requested: 3.75G, maximum: 3.75GB (where the maximum is e.g. 3.749G).

File: batch2/batch/front_end/front_end.py [hunk 14]
```diff
@@ -283,7 +285,22 @@
...
+                    raise web.HTTPBadRequest(
+                        reason=f'resource requests for job {id} are unsatisfiable: '
+                        f'requested: cpu={resources["cpu"]}, memory={resources["memory"]} '
+                        f'maximum: cpu={WORKER_CORES}, memory={total_memory_available}GB')
```
Comment (cseed) [inline batch2/batch/front_end/front_end.py]: Leave off the B.  We should use the same syntax as our resource requests.

File: batch2/batch/utils.py [hunk 15]
```diff
@@ -17,6 +19,53 @@ def parse_cpu_in_mcpu(cpu_string):
...
+        m = 6.5
+    else:
+        assert worker_type == 'highcpu', worker_type
+        m = 0.9
+    return m
+
+
+def worker_memory_per_core_bytes(worker_type):
+    m = worker_memory_per_core_gb(worker_type)
+    return math.ceil(m * 1000**3)  # GCE memory/core are in GB not GiB
```
Comment (cseed) [inline batch2/batch/utils.py]: Hmm, no ceil here, I think.  You can't use a fraction of a byte.

File: batch2/batch/utils.py [hunk 16]
```diff
@@ -17,6 +19,53 @@ def parse_cpu_in_mcpu(cpu_string):
...
+
+
+def memory_bytes_to_cores_mcpu(memory_in_bytes, worker_type):
+    return math.ceil((memory_in_bytes / worker_memory_per_core_bytes(worker_type)) * 1000)
```
Comment (cseed) [inline batch2/batch/utils.py]: This was the ceil I asked for.  The thing we want is:

 > cores_mcpu_to_memory_bytes(memory_bytes_to_cores_mcpu(M, T), T) >= M

File: batch2/batch/utils.py [hunk 17]
```diff
@@ -17,6 +19,53 @@ def parse_cpu_in_mcpu(cpu_string):
...
+
+
+def cores_mcpu_to_memory_bytes(cores_in_mcpu, worker_type):
+    return math.ceil((cores_in_mcpu / 1000) * worker_memory_per_core_bytes(worker_type))
```
Comment (cseed) [inline batch2/batch/utils.py]: No ceil here, I think.  Again, you can't use a fractional byte.

File: batch2/batch/front_end/front_end.py [hunk 18]
```diff
@@ -283,7 +285,22 @@
                 if 'memory' not in resources:
                     resources['memory'] = BATCH_JOB_DEFAULT_MEMORY
 
-                cores_mcpu = parse_cpu_in_mcpu(resources['cpu'])
+                req_cores_mcpu = parse_cpu_in_mcpu(resources['cpu'])
+                req_memory_bytes = parse_memory_in_bytes(resources['memory'])
+
+                if req_cores_mcpu == 0:
+                    raise web.HTTPBadRequest(
+                        reason=f'bad resource request for job {id}: '
```
Comment (cseed) [inline batch2/batch/front_end/front_end.py]: This needs to be "bad resource reqests" (note the s) or, the test for this messages needs to have the s removed.

**Pull request merged**